File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Checks if the lasagna is done. Return the cooking status.
 */
export function cookingStatus(remainingTime) {
  
  if (remainingTime === undefined) return 'You forgot to set the timer.'; 
  
  if (remainingTime <= 0) return 'Lasagna is done.';
  
  return 'Not done, please wait.';
}

/**
 * Computes the preparation time for the Lasagna based on the number of layers.
 * If preparation time is not provided, assume 2mn per layer.
 */
export function preparationTime(layers, timePerLayer = 2) {

  return layers.length * timePerLayer;
}

/**
 * Computes the quantities of noodles and sauce required for a specific recipe.
 */
export function quantities(layers) {
  const result = {noodles: 0, sauce: 0};
  for (const layer of layers) {
    switch (layer) {
      case 'sauce':
        result.sauce += 0.2;
        break;
      case 'noodles':
        result.noodles += 50;
        break;
    }
  }
  return result;
}

/**
 * Adds a secret ingredient (last item in the first recipe) to your recipe (the second recipe).
 * The function doesn't return any result (undefined). The first recipe is unchanged.
 * The second recipe has the secret incredient added to the end.
 */
export function addSecretIngredient(friendList, myList) {
  const secretIngredient = friendList[friendList.length-1];
  myList.push(secretIngredient);
}

/**
 * Scales the recipe from 2 portions to the specified number of portions.
 * Each item in the recipe is scaled approprietly.
 */
export function scaleRecipe(recipe, portions) {
  const scale = portions / 2;
  const scaledRecipe = {};
  for (let ingredient in recipe) {
    scaledRecipe[ingredient] = recipe[ingredient] * scale;
  }
  return scaledRecipe;
}

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

const noodleQtyPerLayer = 50
const sauceQtyPerLayer = 0.2

// PreparationTime computes the time required to prepare the lasagna based
// on the number of layers and the preparation time per layer.
func PreparationTime(layers []string, layerPrepTime int) int {
    // If layerPrepTime is invalid, set to 2.
    if layerPrepTime == 0 {
        layerPrepTime = 2
    }
    return len(layers) * layerPrepTime
}

// Quantities computes the quantity of noodles and sauce required
// to make the lasagna as a function of the number of layers.
func Quantities(layers []string) (noodleQty int, sauceQty float64) {
    for _, layer := range layers {
        if layer == "noodles" {
            noodleQty += noodleQtyPerLayer
        } else if layer == "sauce" {
            sauceQty += sauceQtyPerLayer
        }
    }
    return
}

// AddSecretIngredient adds your friend secret ingredient to your list of
// ingredients by replacing the last ingredient in your list by the last
// ingredient in his list.
func AddSecretIngredient(friendList, myList []string) {
    myList[len(myList)-1] = friendList[len(friendList)-1]
}

// ScaleRecipe scales the quantities from the original recipe (2 portions) to 
// the specified number of portions.
func ScaleRecipe(quantities []float64, numberOfPortions int) []float64 {
    scaleFactor := float64(numberOfPortions) / 2
    scaledQuantities := make([]float64, len(quantities))
    for i := 0; i < len(quantities); i++ {
        scaledQuantities[i] = scaleFactor * quantities[i]
    }
    return scaledQuantities
}


